File: ios/RecipeClipper/UI/Edit/EditRecipeViewModel.swift
import Foundation
import Observation

/// `isNew` is "New recipe" from Home; otherwise the recipe `draft` was loaded from.
/// `showInvalid` is set by a Save that failed validation, so the rule is only pointed out once
/// the user has tried. `savedId` is set once the save lands, for the view to open it.
struct EditRecipeUiState: Equatable {
    let isNew: Bool
    var loading: Bool
    var draft = RecipeDraft()
    var showInvalid = false
    var saving = false
    var saveFailed = false
    /// The recipe to edit is gone (deleted elsewhere).
    var missing = false
    var savedId: Int64?
    /// A new recipe couldn't be saved: the free library is full and all protected (#107).
    var libraryFull = false
    /// A purchase from that prompt that is pending or failed; shown until the next edit.
    var unlockNotice: PurchaseOutcome?

    init(isNew: Bool) {
        self.isNew = isNew
        loading = !isNew
    }
}

/// Edits a recipe's content, or types a new one in (#29; Android's EditRecipeViewModel).
/// Opened with a recipe id from the recipe screen's overflow menu, or with none from Home's
/// "New recipe".
@MainActor
@Observable
final class EditRecipeViewModel {
    private(set) var uiState: EditRecipeUiState
    @ObservationIgnored private let recipeId: Int64?
    @ObservationIgnored private let repository: RecipeRepository
    @ObservationIgnored private let entitlements: Entitlements

    init(recipeId: Int64?, repository: RecipeRepository, entitlements: Entitlements = UnavailableEntitlements()) {
        self.entitlements = entitlements
        self.recipeId = recipeId.flatMap { $0 > 0 ? $0 : nil }
        self.repository = repository
        uiState = EditRecipeUiState(isNew: self.recipeId == nil)
        if let id = self.recipeId {
            Task { [weak self, repository] in
                let recipe = await repository.open(id: id)
                guard let self else { return }
                uiState.loading = false
                if let recipe {
                    uiState.draft = RecipeDraft.of(recipe)
                } else {
                    uiState.missing = true
                }
            }
        }
    }

    func onDraftChange(_ draft: RecipeDraft) {
        uiState.draft = draft
        uiState.saveFailed = false
        uiState.unlockNotice = nil
    }

    /// Unlock from the full-library prompt (#107), then save the recipe as typed.
    func onUnlock() {
        uiState.libraryFull = false
        Task { [weak self, entitlements] in
            let outcome = await entitlements.purchase()
            guard let self else { return }
            if outcome == .unlocked {
                onSave()
            } else if outcome.needsNotice {
                uiState.unlockNotice = outcome
            }
        }
    }

    func onLibraryFullDismiss() { uiState.libraryFull = false }

    func onSave() {
        let state = uiState
        guard !state.loading, !state.saving, !state.missing else { return }
        guard state.draft.isValid else {
            uiState.showInvalid = true
            return
        }
        uiState.saving = true
        uiState.saveFailed = false
        Task { [weak self, recipeId, repository] in
            let saved: Recipe?
            if let recipeId {
                saved = await repository.saveEdit(id: recipeId, draft: state.draft)
            } else {
                saved = await repository.addManual(draft: state.draft)
            }
            guard let self else { return }
            uiState.saving = false
            if let saved, saved.id == 0 {
                uiState.libraryFull = true
            } else if let saved {
                uiState.savedId = saved.id
            } else {
                uiState.saveFailed = true
            }
        }
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/edit/EditRecipeViewModel.kt
package com.example.recipeclipper.ui.edit

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.recipeclipper.data.Entitlements
import com.example.recipeclipper.data.PurchaseOutcome
import com.example.recipeclipper.data.RecipeRepository
import com.example.recipeclipper.data.needsNotice
import com.example.recipeclipper.data.model.RecipeDraft
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * [isNew] is "New recipe" from Home; otherwise the recipe [draft] was loaded from.
 * [showInvalid] is set by a Save that failed validation, so the rule is only pointed out
 * once the user has tried. [savedId] is set once the save lands, for the screen to open it.
 */
data class EditRecipeUiState(
    val isNew: Boolean,
    val loading: Boolean = !isNew,
    val draft: RecipeDraft = RecipeDraft(),
    val showInvalid: Boolean = false,
    val saving: Boolean = false,
    val saveFailed: Boolean = false,
    /** The recipe to edit is gone (deleted elsewhere). */
    val missing: Boolean = false,
    val savedId: Long? = null,
    /** A new recipe couldn't be saved: the free library is full and all protected (#107). */
    val libraryFull: Boolean = false,
    /** A purchase from that prompt that is pending or failed; shown until the next edit. */
    val unlockNotice: PurchaseOutcome? = null
)

/**
 * Edits a recipe's content, or types a new one in (#29). Opened with a recipe id from the
 * recipe screen's overflow menu, or with none from Home's "New recipe".
 */
@HiltViewModel
class EditRecipeViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val repository: RecipeRepository,
    private val entitlements: Entitlements = Entitlements.Unavailable
) : ViewModel() {

    private val recipeId: Long? = savedStateHandle.get<Long>(RECIPE_ID_ARG)?.takeIf { it > 0 }

    private val _uiState = MutableStateFlow(EditRecipeUiState(isNew = recipeId == null))
    val uiState: StateFlow<EditRecipeUiState> = _uiState.asStateFlow()

    init {
        if (recipeId != null) {
            viewModelScope.launch {
                val recipe = repository.open(recipeId)
                _uiState.update {
                    if (recipe == null) it.copy(loading = false, missing = true)
                    else it.copy(loading = false, draft = RecipeDraft.of(recipe))
                }
            }
        }
    }

    fun onDraftChange(draft: RecipeDraft) =
        _uiState.update { it.copy(draft = draft, saveFailed = false, unlockNotice = null) }

    fun onSave() {
        val state = _uiState.value
        if (state.loading || state.saving || state.missing) return
        if (!state.draft.isValid) {
            _uiState.update { it.copy(showInvalid = true) }
            return
        }
        _uiState.update { it.copy(saving = true, saveFailed = false) }
        viewModelScope.launch {
            val saved = if (recipeId == null) {
                repository.addManual(state.draft)
            } else {
                repository.saveEdit(recipeId, state.draft)
            }
            _uiState.update {
                when {
                    saved == null -> it.copy(saving = false, saveFailed = true)
                    saved.id == 0L -> it.copy(saving = false, libraryFull = true)
                    else -> it.copy(saving = false, savedId = saved.id)
                }
            }
        }
    }

    /** Unlock from the full-library prompt (#107), then save the recipe as typed. */
    fun onUnlock() {
        _uiState.update { it.copy(libraryFull = false) }
        viewModelScope.launch {
            val outcome = entitlements.purchase()
            if (outcome == PurchaseOutcome.UNLOCKED) onSave()
            else if (outcome.needsNotice) _uiState.update { it.copy(unlockNotice = outcome) }
        }
    }

    fun onLibraryFullDismiss() = _uiState.update { it.copy(libraryFull = false) }

    companion object {
        const val RECIPE_ID_ARG = "recipeId"
    }
}
